File: server/src/ai/graph.ts
import { StateGraph, END, START } from "@langchain/langgraph";
import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
import { SystemMessage, HumanMessage, BaseMessage } from "@langchain/core/messages";
import * as dotenv from "dotenv";

dotenv.config();

// Define the state type for the graph
interface AgentState {
  messages: BaseMessage[];
  language: string;
}

const llm = new ChatGoogleGenerativeAI({
  modelName: "gemini-1.5-flash",
  maxOutputTokens: 2048,
});

async function businessAgentNode(state: AgentState): Promise<Partial<AgentState>> {
  const { messages, language } = state;
  const systemPrompt = new SystemMessage(
    `You are the EmpowerHub AI Business Operating System. 
    You must strictly respond in the user's preferred language: ${language}.
    Your goal is to orchestrate business creation, marketing, and education.`
  );
  
  const response = await llm.invoke([systemPrompt, ...messages]);
  return { messages: [response] };
}

// Define the graph
const workflow = new StateGraph<AgentState>({
  channels: {
    messages: {
      value: (x: BaseMessage[], y: BaseMessage[]) => x.concat(y),
      default: () => [],
    },
    language: {
      value: (x: string, y: string) => y ?? x,
      default: () => "English",
    }
  }
});

workflow.addNode("agent", businessAgentNode);
workflow.addEdge(START, "agent");
workflow.addEdge("agent", END);

export const app = workflow.compile();

File: ai_service/graph.py
from typing import Annotated, TypedDict
from langgraph.graph import StateGraph, START, END
from langchain_openai import ChatOpenAI
from langchain_core.messages import SystemMessage, HumanMessage, AnyMessage
from langgraph.graph.message import add_messages

class AgentState(TypedDict):
    messages: Annotated[list[AnyMessage], add_messages]
    language: str

def get_llm():
    return ChatOpenAI(model="gpt-4o")

def business_agent(state: AgentState):
    llm = get_llm()
    system_msg = SystemMessage(
        content=f"You are the Empower Hub AI Business Operating System. "
                f"You are talking to a user whose preferred language is {state.get('language', 'English')}. "
                f"Your goal is to guide them from zero knowledge to a successful business."
    )
    messages = [system_msg] + state["messages"]
    response = llm.invoke(messages)
    return {"messages": [response]}

# Define the graph
workflow = StateGraph(AgentState)
workflow.add_node("agent", business_agent)
workflow.add_edge(START, "agent")
workflow.add_edge("agent", END)
app = workflow.compile()
